### ISSUE TITLE
`RemoteClient` receives biased towards early connections.

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#9196]
The `SelectAll<SplitStream<FramedClient<TcpStream>>>` type prioritizes connections early in the sequence, and can starve later connections if the first are kept full.

**Comment by benesch** [comment]
Hmm, where are you seeing that? It's a bit hard to follow the impl of `SelectAll` through `FuturesUnordered`, but at some point in the past I had to add `select_all_biased` (https://github.com/MaterializeInc/materialize/commit/6fc5d531084ebb6d04a4a98bef7e2cb48941c813), which implies that `select_all` either is or was unbiased at some point.

**Comment by benesch** [comment]
If we're worried about `SelectAll`, [`StreamMap`](https://docs.rs/tokio-stream/0.1.8/tokio_stream/struct.StreamMap.html) is the equivalent type from Tokio and its implementation picks a random stream to start polling first.

**Comment by frankmcsherry** [comment]
It looks like they are round robin, not biased. But, the docs implied (to me) that they are biased: "The stream will yield items as they become available on the underlying streams internally, in the order they become available.". But, building a `SelectAll` by pushing a sequence of unbounded streams actually does results in their interleaving, rather than draining in order, which isn't what I would expect from the docs (maybe having expectations is the bug).

Tbh, the thing I'm worried about is control, more than bias. You don't always want round robin, and we don't seem to get to choose. The common flow-control thing I'm used to, for example, is to pinch each of the queues when observed, and drain each pinched queue fully.

**Comment by benesch** [comment]
> Tbh, the thing I'm worried about is control, more than bias. You don't always want round robin, and we don't seem to get to choose. The common flow-control thing I'm used to, for example, is to pinch each of the queues when observed, and drain each pinched queue fully.

Right, yeah, if we want a different control flow mechanism we'd have to write our own version of `SelectAll`. I think a combinator that does the thing you suggest is... dare I say it... relatively straightforward?

```rust
pub struct SelectAll<S> {
    streams: Vec<S>>,
    index: usize,
}

impl<S> Stream for SelectAll<S>
where
    S: Stream + Unpin,
{
    type Item = S::Item;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        for _ in (0..self.streams.len()) {
            match self.streams[self.index].poll_next(cx) {
                Poll::Ready(Some(item)) => return Poll::Ready(Some(item)),
                Poll::Ready(None) => self.streams.remove(self.index),
                Poll::Pending => self.index += 1,
            }
        }
        Poll::Pending
    }
}
```

**Comment by benesch** [comment]
Ah, but that would rely on someone draining the `SelectAll` into a buffer as quickly as possible before proceeding. We'd probably want to do that buffering internally. Maybe something like this?

```rust
pub struct SelectAll<S> {
    streams: Vec<S>>,
    buffer: VecDeque<S::Item>,
    index: usize,
}

impl<S> Stream for SelectAll<S>
where
    S: Stream + Unpin,
{
    type Item = S::Item;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        if let Some(item) = self.buffer.pop_front() {
            return Poll::Ready(Some(item));
        }
        for _ in (0..self.streams.len()) {
            loop {
                match self.streams[self.index].poll_next(cx) {
                    Poll::Ready(Some(item)) => self.buffer.push_back(item),
                    Poll::Ready(None) => {
                        self.streams.remove(self.index);
                        break;
                    }
                    Poll::Pending => {
                        self.index += 1;
                        break;
                    }
                }
            }
        }
        match self.buffer.pop_front() {
            Some(item) => Poll::Ready(Some(item)),
            None => Poll::Pending,
        }
    }
}
```
